### ISSUE TITLE
type inference failed for scala 2.10

### ISSUE THREAD

**Issue by manuzhang** [gearpump/gearpump#1304]
```bash
[error] /Users/doria/git/gearpump/examples/streaming/wordcount/src/main/scala/org/apache/gearpump/streaming/examples/wordcount/WordCount.scala:44: inferred type arguments [org.apache.gearpump.streaming.Processor.DefaultProcessor[_ >: org.apache.gearpump.streaming.examples.wordcount.Split with org.apache.gearpump.streaming.examples.wordcount.Sum <: org.apache.gearpump.streaming.task.Task],E] do not conform to method apply's type parameter bounds [T <: org.apache.gearpump.streaming.Processor[org.apache.gearpump.streaming.task.Task],P <: org.apache.gearpump.partitioner.Partitioner]
[error]     val app = StreamApplication("wordCount", Graph(split ~> sum), UserConfig.empty)
[error]               ^
[error] /Users/doria/git/gearpump/examples/streaming/wordcount/src/main/scala/org/apache/gearpump/streaming/examples/wordcount/WordCount.scala:44: type mismatch;
[error]  found   : org.apache.gearpump.util.Graph[org.apache.gearpump.streaming.Processor.DefaultProcessor[_ >: org.apache.gearpump.streaming.examples.wordcount.Split with org.apache.gearpump.streaming.examples.wordcount.Sum <: org.apache.gearpump.streaming.task.Task],E]
[error]  required: org.apache.gearpump.util.Graph[T,P]
[error]     val app = StreamApplication("wordCount", Graph(split ~> sum), UserConfig.empty)
[error]                                                   ^
[error] two errors found
```

adding partitioner could fix the issue

```scala
val app = StreamApplication("wordcount", Graph(split ~ new HashPartitioner ~> sum), UserConfig.empty)
```

**Comment by manuzhang** [comment]
this won't compile even with `partitioner`

```scala
    StreamApplication("transport", Graph(source ~ partitioner ~> inspector, queryServer), userConfig)
```

error info
```bash
[error] /Users/doria/git/gearpump/examples/streaming/stockcrawler/src/main/scala/org/apache/gearpump/streaming/examples/stock/main/Stock.scala:60: type mismatch;
[error]  found   : org.apache.gearpump.util.Graph[org.apache.gearpump.streaming.Processor.DefaultProcessor[_ >: org.apache.gearpump.streaming.examples.stock.QueryServer with org.apache.gearpump.streaming.examples.stock.Crawler with org.apache.gearpump.streaming.examples.stock.Analyzer <: org.apache.gearpump.streaming.task.Task],Any]
[error]  required: org.apache.gearpump.util.Graph[T,P]
[error]       Graph(crawler ~ partitioner ~> analyzer, queryServer), userConfig
[error] 
```

**Comment by kkasravi** [comment]
What happens if you provide a type definition within Stock.scala?
```
type T = org.apache.gearpump.streaming.examples.stock.QueryServer with org.apache.gearpump.streaming.examples.stock.Crawler with org.apache.gearpump.streaming.examples.stock.Analyzer
```
and explicit typing when calling Graph?
```
Graph[T <: Task, Any](crawler, queryServer)
```

**Comment by clockfly** [comment]
I don't know why the type inference is not working. 

As a work-around, we can do type cast in StreamApplication.

**Comment by clockfly** [comment]
@manuzhang 

Can you fix this by adding type cast, and runtime type check?

**Comment by manuzhang** [comment]
@kkasravi @clockfly  I'll try your suggestions

**Comment by manuzhang** [comment]
fixed in #1311
